core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,177 @@ angular.module('oppia').factory('ImprovementModalService', [
+                MAX_UNRELATED_ACTIONS_PER_BLOCK) {
+              $scope.playthroughIndex = playthroughIndex;
+
+              // This is the index of the first action in the range of actions
+              // displayed. Initially, only one action is displayed. When this
+              // index reaches 0, we've displayed all the learner actions.
+              let indexOfFirstDisplayedAction = playthrough.actions.length - 1;
+
+              /**
+               * Identifies whether all learner actions have been displayed and
+               * no further expansion is possible.
+               * @returns {boolean}
+               */
+              $scope.canExpandActions = function() {
+                return indexOfFirstDisplayedAction !== 0;
+              };
+
+              /**
+               * Iterates through the learner actions and expands the set of
+               * actions to be displayed. If all learner actions have been
+               * handled, no more expansion is possible.
+               */
+              let expandActions = function() {
+                let i;
+                let previousStateName = null;
+                if (indexOfFirstDisplayedAction <
+                    playthrough.actions.length - 1) {
+                  let action = playthrough.actions[
+                    indexOfFirstDisplayedAction];
+                  previousStateName =
+                      action.actionCustomizationArgs.state_name.value;
+                }
+                for (i = indexOfFirstDisplayedAction; i >= 0; i--) {
+                  let action = playthrough.actions[i];
+                  let currentStateName =
+                      action.actionCustomizationArgs.state_name.value;
+
+                  if (currentStateName !== previousStateName) {
+                    // When the latest state name and the state name of the
+                    // current learner action differ, they are part of
+                    // different contexts. So, we identify whether the size
+                    // of the block has crossed the threshold number of actions
+                    // per block. If it has, we don't add more learner actions
+                    // to the block.
+                    if (indexOfFirstDisplayedAction - i >= (
+                      MAX_UNRELATED_ACTIONS_PER_BLOCK)) {
+                      indexOfFirstDisplayedAction = i + 1;
+                      break;
+                    }
+                  }
+                  previousStateName = currentStateName;
+                }
+                // This is the case when all learner actions have been iterated
+                // over, and no more expansion is possible.
+                if (i === -1) {
+                  indexOfFirstDisplayedAction = 0;
+                }
+              };
+
+              expandActions();
+
+              // Index to know where to start highlighting actions.
+              let indexOfActionInFirstDisplayedBlock =
```

Comment:
Still no idea why this needs to be a separate variable. You don't ever set it to anything besides indexOfFirstDisplayedAction...

Reply:
@seanlip But this is set the value of the indexOfFirstDisplayedAction only for the first iteration of expanded actions. Post that, the indexOfFirstDisplayedAction keeps getting updated, but this variable does not.

Reply:
Ohh, I see, thanks for clarifying. Can you call it something like actionIndexAtStartOfFinalBlock or something like that? Add a comment to explain this a bit more too (including what you said above). Let's use chronological order for things like "initial"/"final".

Reply:
Done.